EMI: Implement mesh-specific alpha values

chkr-private opened the pull request:
* if Lua_V2::SetActorGlobalAlpha() is called with a mesh name,
  set a mesh-specific alpha value / mode
* use this alpha value in drawEMIModelFace()
* this commit fixes the problem that the whole actor vanishs in some scenes
   * issue #896: when entering or leaving the porch of the of LUA bar 
   * issue #1151: when jumping out/in the bank window)
* instead of fading the actor in or out, only his shadow is faded correctly
* implemented for all 3 graphic drivers

somaen commented on engines/grim/actor.cpp:
```diff
@@ -1446,6 +1446,48 @@ float Actor::getLookAtRate() const {
+	return nullptr;
+}
+
+void
```
No newline her (nor on the next function.)

somaen commented on engines/grim/emi/costumeemi.cpp:
```diff
@@ -315,4 +315,13 @@ EMIModel *EMICostume::getEMIModel() const {
 }
 
+EMIModel *EMICostume::getEMIModel(int num) const {
+	if (num >= _numChores) return nullptr;
```
newline and indent on these three please.

somaen commented on engines/grim/emi/modelemi.cpp:
```diff
@@ -105,6 +105,12 @@ void EMIModel::loadMesh(Common::SeekableReadStream *data) {
 
 	char f[4];
 	data->read(f, 4);
+	for (uint l = 0; l < nameString.size(); ++l)
```
I'd prefer braces on a non-trivial for-loop like this.

chkr-private commented:
I have addressed all review comments, implemented the feature also for opengls (@Botje: please can you review my changes?), squashed all commits and rebased to latest master.

Botje commented:
@chkr-private, looks good to me. :+1:

klusark commented on engines/grim/actor.h:
```diff
@@ -538,8 +539,8 @@ class Actor : public PoolObject<Actor> {
 	float getEffectiveAlpha() const { return _alphaMode != AlphaOff ? _globalAlpha : 1.f; }
-	void setGlobalAlpha(float alpha) { _globalAlpha = alpha; }
-	void setAlphaMode(AlphaMode mode) { _alphaMode = mode; }
+	void setGlobalAlpha(float alpha, const char* mesh);
```
Can we make these const Common::String? Same with findModelWithMesh.

somaen replied:
Well, preferably const Common::String& I'd assume.

chkr-private replied:
Good point! I changed the function signatures accordingly. Please see the additional commit (should be squashed before the final merge).

somaen commented on engines/grim/emi/lua_v2_actor.cpp:
```diff
@@ -88,13 +88,17 @@ void Lua_V2::SetActorGlobalAlpha() {
 	if (!lua_isnumber(alphaObj))
 		return;
 
+	const char* mesh = nullptr;
```
* aligns to the name, not the type.

somaen commented on engines/grim/actor.h:
```diff
@@ -587,6 +588,7 @@ class Actor : public PoolObject<Actor> {
 	void calculateOrientation(const Math::Vector3d &pos, Math::Angle *pitch, Math::Angle *yaw, Math::Angle *roll);
 
 	bool getSphereInfo(bool adjustZ, float &size, Math::Vector3d &pos) const;
+	EMIModel* findModelWithMesh(const Common::String &mesh);
```
* aligns to the function, not the type.

somaen commented:
Sorry for the late extra comments, but I don't see anything beyond those though. So this can probably be merged sooner rather than later.

chkr-private commented:
@somaen, no problem - thanks for the review. I fixed all reported issues and rebased again.

chkr-private commented:
@klusark: I moved the initialization of _meshAlpha and _meshAlphaMode into the constructor and rebased/squashed the commits.
